src/utils/retry.ts

```diff
@@ -0,0 +1,77 @@
+// Copyright IBM Corp. 2018. All Rights Reserved.
+// Node module: @loopback/example-shopping
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {HttpErrors} from '@loopback/rest';
+import {promisify} from 'util';
+import * as debugFactory from 'debug';
+const debug = debugFactory('loopback:example:shopping');
+
+export interface TaskStatus<T> {
+  done: boolean;
+  value?: T | null;
+}
+
+/**
+ * A task that can be retried
+ */
+export interface Task<T> {
+  run(): Promise<TaskStatus<T>>;
+  description: string;
+}
+
+/**
+ * Options for retry
+ */
+export interface RetryOptions {
+  /**
+   * Maximum number of tries including the first run.
+   */
+  maxTries?: number;
+  /**
+   * Milliseconds to wait after each try
+   */
+  interval?: number;
+}
+
+/**
+ * Retry a task for number of times with the given interval in ms
+ * @param task Task object {run, description}
+ * @param maxTries Maximum number of tries (including the first run),
+ * default to 10
+ * @param interval Milliseconds to wait after each try, default to 100ms
+ */
+export async function retry<T>(
+  task: Task<T>,
+  {maxTries = 10, interval = 100}: RetryOptions = {},
+): Promise<T> {
+  if (maxTries < 1) maxTries = 1;
+  let triesLeft = maxTries;
+  while (true) {
+    debug(
+      'Try %s (%d/%d)',
+      task.description,
+      maxTries - triesLeft + 1,
+      maxTries,
+    );
+    const status = await task.run();
+    if (status.done) return status.value!;
```

Comment:
```ts
status.value!
```

Please don't! The typings are designed in such way that `task.run()` can return a value that's undefined.

If you want to preserve `TaskStatus` interface then you need to change `retry` to return `T | null` and push the responsibility of handling null/undefined result to callers of `retry`. This is not a great UX to me.

I think a better solution is to follow my proposal outlined in https://github.com/strongloop/loopback4-example-shopping/pull/14#discussion_r219851802, I believe it provides information to allow TypeScript understand that when `status.done` is true, then `status.value` must be defined.

```ts
export interface TaskFinished<T> {
  done: true;
  result: T;
}

export interface RetryNeeded {
  done: false;
}

export interface Task<T> {
  run(): Promise<TaskFinished<T> | RetryNeeded>;
  description?: string;
}
```